libs/StateSync.Managed/src/NetworkManager.cs

```diff
@@ -0,0 +1,222 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.MixedReality.Sharing.StateSync
+{
+    public interface NetworkManager
+    {
+        NetworkConnection GetConnection(InternedBlobRef connectionString);
+
+        bool PollMessage(NetworkListener listener);
+    }
+
+    /// <summary>
+    /// Maintains the cache of C++ wrappers for the created channels.
+    /// </summary>
+    public class NetworkManagerPInvokeWrapper
+    {
+        private readonly NetworkManager manager_;
+
+        // A connection created on the C# side can be wrapped by a NetworkConnection object on C++ side,
+        // that will hold a GCHandle to the C# NetworkConnection.
+        //
+        // To avoid the creation of the new wrapper every time the C++ side calls GetConnection,
+        // we are maintaining the cache of already created wrappers, but to avoid cyclic references,
+        // we only store a weak pointer to the wrapper.
+        // The weak pointer is separately allocated on the C++ side, and the pointer to it is
+        // stored as a handle inside NetworkConnectionWeakPtr, witch are values in connectionWrappersCache_
```

Comment:
nit:
```suggestion
        // stored as a handle inside an instance of CppConnectionWeakPtr, which appears as a value in connectionWrappersCache_
```
Rename to CppConnectionWeakPtr so it's a bit clearer that it refers to a C++ weak ptr to a C++ class and has no (direct) relationship with the C# NetworkConnection class.

Reply:
I'll add some clarity to both class names and variable names, marking the C++ stuff with cpp prefix.